### PR TITLE
Croptype

### PR DESCRIPTION
**Pull request by gabrieltseng**

### CODE REVIEW

File: paper_eval.py [hunk 1]
```diff
@@ -38,11 +42,22 @@
...
-argparser.add_argument("--num_workers", type=int, default=4)
+argparser.add_argument("--num_workers", type=int, default=64)
 argparser.add_argument("--wandb", dest="wandb", action="store_true")
 argparser.add_argument("--wandb_org", type=str, default="nasa-harvest")
 argparser.add_argument("--parquet_file", type=str, default="rawts-monthly_calval.parquet")
-argparser.add_argument("--val_samples_file", type=str, default="cropland_test_split_samples.csv")
+
+# argparser.add_argument("--val_samples_file", type=str, default="cropland_spatial_generalization_test_split_samples.csv")
+
+argparser.add_argument("--presto_model_type", type=str, default="presto-ft-ct")
```
Comment (gabrieltseng) [inline paper_eval.py:52]: Is this argument (`presto_model_type`) only used to name the experiment file? if so can it be given a different name; the current name implies it will somehow affect the model

Comment (gabrieltseng) [reply paper_eval.py:52]: Same for other names which don't affect the functionality, e.g. `compositing_window`. 

Also I think the most up to date `main` hasn't been merged into this, since the 10d compositing is also supported but doesn't seem to be here.

File: paper_eval.py [hunk 2]
```diff
@@ -77,79 +103,102 @@
...
-    model = Presto.load_pretrained()
-    best_model_path: Optional[Path] = default_model_path
-else:
-    if path_to_config == "":
-        path_to_config = config_dir / "default.json"
-    model_kwargs = json.load(Path(path_to_config).open("r"))
-    model = Presto.construct(**model_kwargs)
-    best_model_path = None
-model.to(device)
+val_samples_file = f"{task_type}_{test_type}_generalization_test_split_samples.csv"
```
Comment (gabrieltseng) [inline paper_eval.py:118]: This approach requires the script to be run 3 times to get the full results (instead of only once to get all 3 results, which was the case previously). 

Perhaps we can update this so that a single run gets all the data?

File: paper_eval.py [hunk 3]
```diff
@@ -77,79 +103,102 @@
...
+model_path.mkdir(exist_ok=True, parents=True)
+experiment_prefix = f"{presto_model_type}_{compositing_window}_{task_type}_{test_type}_{time_token}"
+finetuned_model_path = model_path / f"{experiment_prefix}.pt"
+results_path = model_logging_dir / f"{experiment_prefix}.csv"
+catboost_model_path = model_logging_dir / f"{experiment_prefix}.cbm"
+
+# check if finetuned model already exists
+logger.info("Checking if the finetuned model exists")
+if os.path.isfile(finetuned_model_path):
+    logger.info("Finetuned model found! Loading...")
```
Comment (gabrieltseng) [inline paper_eval.py:149]: Do we want to do this? I can imagine this introducing lots of unexpected errors since if a finetuned model exists from a previous run, it would automatically affect this run whether or not we want it to

Comment (gabrieltseng) [reply paper_eval.py:149]: Also would this work if there is no finetuned model? I think the script would error out since `finetuned_model` would never be initialized?

Comment (gabrieltseng) [reply paper_eval.py:149]: Also also, how is the model finetuned in this case? I think thats pretty important so it would be good to capture that in this script

File: presto/eval.py [hunk 4]
```diff
@@ -35,15 +35,14 @@
 class Hyperparams:
     lr: float = 2e-5
     max_epochs: int = 100
-    batch_size: int = 64
-    patience: int = 10
-    num_workers: int = 4
+    batch_size: int = 2048
```
Comment (gabrieltseng) [inline presto/eval.py:38]: How were these chosen? This batch size is probably too large for finetuning

File: presto/eval.py [hunk 5]
```diff
@@ -35,15 +35,14 @@
...
+    patience: int = 3
+    num_workers: int = 8
 
 
 class WorldCerealEval:
-    name = "WorldCerealCropland"
```
Comment (gabrieltseng) [inline presto/eval.py:44]: We probably want to keep the Cropland code too, right? Or is it being completely replaced with the crop type code?

File: presto/presto.py [hunk 6]
```diff
@@ -458,12 +486,19 @@ def forward(
 
         # append latlon tokens
         latlon_tokens = self.latlon_embed(self.cartesian(latlons)).unsqueeze(1)
-        x = torch.cat((latlon_tokens, x), dim=1)
-        upd_mask = torch.cat((torch.zeros(x.shape[0])[:, None].to(device), upd_mask), dim=1)
-        orig_indices = torch.cat(
-            (torch.zeros(x.shape[0])[:, None].to(device).int(), orig_indices + 1),
-            dim=1,
-        )
+        x, upd_mask, orig_indices = self.add_token(latlon_tokens, x, upd_mask, orig_indices)
```
Comment (gabrieltseng) [inline presto/presto.py:489]: @cbutsko a hackey way to get the model to ignore latlons is to change the mask here. A mask value of 0s tells the model to include the value, and a mask value of 1s tells it to ignore the value.

New tokens get added to the front of the sequence. Concretely:

`x` has shape `[batch_size, num_tokens, dim]`
and `mask` has shape `[batch_size, num_tokens]`.

The latlons just got added to the front of the sequence, so you can do

```python
upd_mask[:, 0] = 1
```
right after line 489 (where `self.add_token(latlon_tokens, x, upd_mask, orig_indices)` is called) to update the mask so that the latlon token will be ignored entirely.

File: paper_eval.py (hunk 1 above)
Comment (cbutsko) [reply paper_eval.py]: good point. addressed that here https://github.com/WorldCereal/presto-worldcereal/commit/1601dbc3c6cb601846c65e3b18914fb46b5946c0

File: paper_eval.py (hunk 3 above)
Comment (cbutsko) [reply paper_eval.py:177]: I still feel the need for being able to upload the finetuned model, particularly if I want to run downstream classifiers only and collect metrics. in particular, [this piece](https://github.com/WorldCereal/presto-worldcereal/blob/5f5d296962fb158f6f4c9f84e74e0deff2efcde7/paper_eval.py#L173-L200) only checks if the model exists, collects metrics for the uploaded model along with spatial plots and runs sklearn models. 
I didn't succeed in implementing the upload in presto.py, so now this piece wouldn't run anyway. for now, I
just put a placeholder here https://github.com/WorldCereal/presto-worldcereal/commit/532ec67b5a81b0845f05534864550de073957fb3

File: presto/eval.py (hunk 4 above)
Comment (cbutsko) [reply presto/eval.py]: indeed, reverted batch_size to 256 here https://github.com/WorldCereal/presto-worldcereal/commit/a350280be17c7715d4022ea78cd9f478e61bfa3c
I was playing with leaning rate, but also ended up using your value.

File: presto/eval.py (hunk 5 above)
Comment (cbutsko) [reply presto/eval.py:44]: addressed it here https://github.com/WorldCereal/presto-worldcereal/commit/a350280be17c7715d4022ea78cd9f478e61bfa3c

File: presto/presto.py (hunk 6 above)
Comment (cbutsko) [reply presto/presto.py:489]: added this line as you suggested instead of filling latlons with zeros in the dataset
https://github.com/WorldCereal/presto-worldcereal/commit/0cf1e6072e129499fcc205ee11d3b7dc43dad472

**Comment by gabrieltseng** [comment]
- @cbutsko to fix the tests (code is already working) - ETA end of week
- @gabrieltseng to take a look / leave comments - early next week
- @gabrieltseng & @cbutsko to discuss / iron things out - late next week
- :tada: merge!

File: .gitignore [hunk 7]
```diff
@@ -20,4 +20,6 @@ gha-creds-*.json
 .idea
 scrap
 output/*
-imgs/*
\ No newline at end of file
+imgs/*
+# don't track catboost training info
+*/catboost_info
```
Comment (gabrieltseng) [inline .gitignore:25]: `catboost_info` looks like it is in git - should that folder be removed from git?

File: presto/eval.py [hunk 8]
```diff
@@ -60,23 +60,74 @@ def __init__(
...
 
         train_data, val_data = WorldCerealLabelledDataset.split_df(train_data, val_size=val_size)
         self.train_df = prep_dataframe(train_data, filter_function, dekadal=dekadal)
         self.val_df = prep_dataframe(val_data, filter_function, dekadal=dekadal)
         self.test_df = prep_dataframe(test_data, filter_function, dekadal=dekadal)
 
+        if task_type == "cropland":
+            self.num_outputs = 1
+            self.croptype_list = []
+        if task_type == "croptype":
```
Comment (gabrieltseng) [inline presto/eval.py:86]: nit: elif

File: presto/presto.py [hunk 9]
```diff
@@ -245,10 +249,11 @@ def get_month_encoding_table(d_hid):
...
 
-    if torch.cuda.is_available():
-        return torch.FloatTensor(month_table).cuda()
-    else:
-        return torch.FloatTensor(month_table)
+    return torch.FloatTensor(month_table).to(device)
+    # if torch.cuda.is_available():
+    #     return torch.FloatTensor(month_table).cuda()
+    # else:
+    #     return torch.FloatTensor(month_table)
```
Comment (gabrieltseng) [inline presto/presto.py:256]: Nit: these commented lines can be removed

File: presto/presto.py [hunk 10]
```diff
@@ -185,7 +187,8 @@ def __init__(
         norm_layer=nn.LayerNorm,
     ):
         super().__init__()
-        self.norm1 = norm_layer(dim)
+        self.norm1 = norm_layer(dim, device=device)
+        # self.norm1 = norm_layer(dim)
```
Comment (gabrieltseng) [inline presto/presto.py:191]: Why are these `to_device` calls necessary? They should all be handled when you call `model.to(device)`

File: .gitignore (hunk 7 above)
Comment (kvantricht) [reply .gitignore:25]: It should indeed not be in git.

File: presto/dataset.py [hunk 11]
```diff
@@ -192,20 +258,31 @@ def split_df(
...
 
 class WorldCerealMaskedDataset(WorldCerealBase):
-    def __init__(self, dataframe: pd.DataFrame, mask_params: MaskParamsNoDw):
+    def __init__(
+        self,
+        dataframe: pd.DataFrame,
+        mask_params: MaskParamsNoDw,
+        task_type: str = "cropland",
+        croptype_list: List = [],
+        model_mode: str = "",
```
Comment (gabrieltseng) [inline presto/dataset.py:272]: Am I right in understanding that `model_mode` only takes two values - `"Hierarchical CatBoostClassifier"` or `""`?

If so can this be replaced with a boolean? e.g. `return_hierarchical_labels: bool = False`? This would make it clearer what this parameter controls

File: presto/dataset.py [hunk 12]
```diff
@@ -61,21 +65,50 @@ def __len__(self):
         return self.df.shape[0]
 
     @staticmethod
-    def target_crop(row_d: Dict) -> int:
-        # by default, we predict crop vs non crop
-        return int(row_d["LANDCOVER_LABEL"] == 11)
+    def target_crop(
```
Comment (gabrieltseng) [inline presto/dataset.py:68]: nit: this probably makes sense to move into the `WorldCerealLabelledDataset` since its now tightly coupled to attributes in the labelled dataset

**Comment by kvantricht** [comment]
Before merging this one to `main`, please bump the version number [here](https://github.com/WorldCereal/presto-worldcereal/blob/croptype/setup.py#L35) to avoid breakage in inference.

**Comment by kvantricht** [comment]
https://github.com/WorldCereal/presto-worldcereal/blob/croptype/setup.py#L35

**Comment by kvantricht** [comment]
Before merging this PR, please bump version number [here](https://github.com/WorldCereal/presto-worldcereal/blob/croptype/setup.py#L35) to avoid breaking the inference.

File: presto/dataset.py (hunk 11 above)
Comment (cbutsko) [reply presto/dataset.py:272]: hmmm, good idea 🤔 
my thinking was that `model_mode` parameter already existed, and thus can be just passed down from the the very top... 

I use it here when I iterate over the possible model_modes (`Regression, Random Forest, CatBoostClassifier, Hierarchical CatBoostClassifier`), and pass this parameter to `DataLoader` so that I can get correct targets from it. 
https://github.com/WorldCereal/presto-worldcereal/blob/d72d8199d07a50bcfc2f01075ef75f5229f16bc6/presto/eval.py#L231

If you can suggest more elegant way to do it without dragging this `model_mode` that far, I'd be happy to try it. I don't it see how to overcome it now 🙈

File: presto/dataset.py (hunk 12 above)
Comment (cbutsko) [reply presto/dataset.py]: not really sure how to do that correctly... this method is used in `row_to_arrays`, which is defined in `WorldCerealBase`, and then inherited by all the child classes...
so moving the `target_crop` away would require to rework this `row_to_arrays` too, right?

Comment (kvantricht) [reply presto/dataset.py]: I guess indeed it makes sense to remove `target_function` from `row_to_arrays` in order for the latter to focus only on the input data for presto. The return value of that method than no longer includes the target. In subclasses that need a target (maybe even only `WorldCerealLabelledDataset`) you'd call `row_to_arrays` to collect the inputs and then use `target_function` to assemble the target. The added value here is also that in the case we don't have a ground truth for a row, we don't have to put in hacks to make it work.

File: presto/dataset.py (hunk 11 above)
Comment (gabrieltseng) [reply presto/dataset.py:272]: Potential solution - in the dataset init, have the argument:

```python
    return_hierarchical_labels: bool = False
```
and then when you call the dataset:

```python
   DataLoader(..., return_hierarchical_labels=True if "Hierarchical" in model_mode else False)
```

File: presto/dataset.py (hunk 12 above)
Comment (gabrieltseng) [reply presto/dataset.py]: The labelled dataset `__getitem__` function goes from:

```python
    def __getitem__(self, idx):
        # Get the sample
        df_index = self.indices[idx]
        row = self.df.iloc[df_index, :]
        eo, mask_per_token, latlon, month, valid_month, target = self.row_to_arrays(
            row, self.task_type, self.croptype_list, self.model_mode
        )
        mask_per_variable = np.repeat(mask_per_token, BAND_EXPANSION, axis=1)
        return (
            self.normalize_and_mask(eo),
            target,
            np.ones(self.NUM_TIMESTEPS) * (DynamicWorld2020_2021.class_amount),
            latlon,
            month,
            valid_month,
            mask_per_variable,
```

Now, rows_to_arrays no longer returns a target, so __getitem__ neds to get it:

```python
    def __getitem__(self, idx):
        # Get the sample
        df_index = self.indices[idx]
        row = self.df.iloc[df_index, :]
        eo, mask_per_token, latlon, month, valid_month = self.row_to_arrays(
            row, self.task_type, self.croptype_list, self.model_mode
        )
        target = self.target_function(...)  # new
        mask_per_variable = np.repeat(mask_per_token, BAND_EXPANSION, axis=1)
        return (
            self.normalize_and_mask(eo),
            target,
            np.ones(self.NUM_TIMESTEPS) * (DynamicWorld2020_2021.class_amount),
            latlon,
            month,
            valid_month,
            mask_per_variable,
```

**Comment by gabrieltseng** [comment]
@gabrieltseng to run this branch

File: presto/dataset.py [hunk 13]
```diff
@@ -237,16 +290,32 @@ def __getitem__(self, idx):
...
+        6212,  # temporary crops, too generic
+        7920,
+        9520,
+        3400,
+        3900,  # generic and other classes
+        4390,
+        4000,
+        4300,  # generic and other classes
+    ]
+    df = df[(df["LANDCOVER_LABEL"] == 11) & (~df["CROPTYPE_LABEL"].isin(labels_to_exclude))]
```
Comment (kvantricht) [inline presto/dataset.py:416]: where does this list come from? And should we focus only on `LANDCOVER_LABEL == 11`?
These kind of methods have a pretty important impact downstream, so let's discuss them in depth to determine training strategy.

File: presto/dataset.py [hunk 14]
```diff
@@ -237,16 +290,32 @@ def __getitem__(self, idx):
...
 
 
-def target_maize(row_d) -> int:
-    # 1200 is maize
-    return int(row_d["CROPTYPE_LABEL"] == 1200)
-
-
 class WorldCerealLabelledDataset(WorldCerealBase):
     # 0: no information, 10: could be both annual or perennial
     FILTER_LABELS = [0, 10]
```
Comment (kvantricht) [inline presto/dataset.py:423]: TBD!

File: presto/dataset.py [hunk 15]
```diff
@@ -287,22 +361,79 @@ def __init__(
...
+                for loc_idx, (_, row) in enumerate(self.df.iterrows()):
+                    target = self.target_crop(
+                        row, self.task_type, self.croptype_list, return_hierarchical_labels
+                    )
+                    if target == 0:
+                        neg_indices.append(loc_idx)
+                    else:
+                        pos_indices.append(loc_idx)
+                if len(pos_indices) > len(neg_indices):
+                    self.indices = (
```
Comment (kvantricht) [inline presto/dataset.py:485]: shouldn't we have a similar option to cap the number of repeats for the cropland task?

File: presto/dataset.py [hunk 16]
```diff
@@ -287,22 +361,79 @@ def __init__(
...
+                                tclass_loc_idx = tclass_loc_idx * (
+                                    optimal_class_size // len(tclass_loc_idx)
+                                )
+                        else:
+                            tclass_loc_idx = tclass_loc_idx * (
+                                optimal_class_size // len(tclass_loc_idx)
+                            )
+
+                    balanced_inds.extend(tclass_loc_idx)
+                self.indices = balanced_inds
```
Comment (kvantricht) [inline presto/dataset.py:529]: can we add a logging here that reports how many samples of each class are being repeated? So we can trace this afterwards and also spot any unexpected behavior more easily.

File: presto/dataset.py [hunk 17]
```diff
@@ -287,22 +361,79 @@ def __init__(
...
+    def target_crop(
+        row_d: pd.Series,
+        task_type: str = "cropland",
+        croptype_list: List = [],
+        return_hierarchical_labels: bool = False,
+    ) -> Union[int, np.ndarray, List]:
+
+        _target: Union[int, np.ndarray, List]
+        if task_type == "cropland":
+            _target = int(row_d["LANDCOVER_LABEL"] == 11)
```
Comment (kvantricht) [inline presto/dataset.py:543]: TBD!

File: presto/dataset.py [hunk 18]
```diff
@@ -551,6 +703,11 @@ def nc_to_arrays(
         flat_latlons = cls._extract_latlons(inarr, epsg)
         months = cls._extract_months(inarr)
 
+        # extract valid_date from ds properties and multiply the was as months array
```
Comment (kvantricht) [inline presto/dataset.py:850]: this comment has a typo running it a bit unclear

File: presto/dataset.py [hunk 19]
```diff
@@ -551,6 +703,11 @@ def nc_to_arrays(
...
+        # add a test to it if there's none
+        valid_month = pd.to_datetime(ds.attrs["valid_time"]).month - 1
+        valid_months = np.full_like(months, valid_month)
+
         if cls.Y not in ds:
```
Comment (kvantricht) [inline presto/dataset.py:855]: Is this still relevant? What does this step do?

File: presto/eval.py [hunk 20]
```diff
@@ -60,24 +62,81 @@ def __init__(
...
         self.train_df = prep_dataframe(train_data, filter_function, dekadal=dekadal)
         self.val_df = prep_dataframe(val_data, filter_function, dekadal=dekadal)
         self.test_df = prep_dataframe(test_data, filter_function, dekadal=dekadal)
 
+        if task_type == "cropland":
+            self.num_outputs = 1
+            self.croptype_list = []
+        elif task_type == "croptype":
+            # compress all classes in train that contain less
+            # than MIN_SAMPLES_PER_CLASS samples into "other"
```
Comment (kvantricht) [inline presto/eval.py:94]: TBD, should we do it like this? We're creating a very diverse "other" class with many small exotic classes. Another option is to ignore these classes during training.

File: presto/eval.py [hunk 21]
```diff
@@ -155,92 +229,207 @@ def dataloader_to_encodings_and_targets(
...
+                # # https://github.com/WorldCereal/wc-classification/blob/
+                # # 4a9a839507d9b4f63c378b3b1d164325cbe843d6/src/
+                # # worldcereal/classification/models.py#L490
+
+                if self.task_type == "cropland":
+                    learning_rate = 0.05
+                    l2_leaf_reg = 3
+                if self.task_type == "croptype":
+                    learning_rate = 0.1
+                    l2_leaf_reg = 30
```
Comment (kvantricht) [inline presto/eval.py:304]: where do these values come from?

File: presto/eval.py [hunk 22]
```diff
@@ -268,140 +458,247 @@ def spatial_inference(
...
         optimizer = AdamW(parameters, lr=hyperparams.lr)
+        scheduler = lr_scheduler.ExponentialLR(optimizer, gamma=0.99)
 
         train_ds = self.ds_class(
             self.train_df,
-            countries_to_remove=self.countries_to_remove,
-            years_to_remove=self.years_to_remove,
-            target_function=self.target_function,
-            balance=True,
+            # countries_to_remove=self.countries_to_remove,
```
Comment (kvantricht) [inline presto/eval.py]: why are these kwargs now commented out?

File: presto/inference.py [hunk 23]
```diff
@@ -84,10 +88,13 @@ def _create_dataloader(
 
     def _create_presto_input(
         self, inarr: xr.DataArray, epsg: int = 4326
-    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
         eo_data, mask = self._ds._extract_eo_data(inarr)
         flat_latlons = self._ds._extract_latlons(inarr, epsg)
         months = self._ds._extract_months(inarr)
+
+        valid_month = pd.to_datetime(inarr.attrs["valid_date"]).month - 1
```
Comment (kvantricht) [inline presto/inference.py:103]: this method will also be used during true inference. What if `valid_date` is not present in the array? Either we should check this and throw a clear error, or we should log this and have a workaround to not feed it (if presto can handle it) or take the middle of the `t` coordinate. TBD.

File: presto/inference.py [hunk 24]
```diff
@@ -255,6 +243,34 @@ def process_parquet(df: pd.DataFrame) -> pd.DataFrame:
         "valid_date",
     ]
 
+    # add dummy value + rename stuff for compatibility with existing functions
+    df["OPTICAL-B8A"] = 65535
+
+    rename_dict = {
```
Comment (kvantricht) [inline presto/inference.py]: This `process_parquet` method needs to be checked again. Though on the one hand that this one was taken out, and other hand if it's still in, normally things like this renaming were no longer necessary as it was already taken care of. TBD.

File: presto/inference.py [hunk 25]
```diff
@@ -275,11 +291,27 @@ def process_parquet(df: pd.DataFrame) -> pd.DataFrame:
...
 
     # Now reassign start_date to the actual subset counted back from end_date
-    df["start_date"] = df["end_date"] - pd.DateOffset(years=1) + pd.DateOffset(days=1)
+    # ignoring mypy error here:
+    # error: No overload variant of "__sub__" of "Series" matches argument type "DateOffset"
+    df["start_date"] = (  # type: ignore
+        df["end_date"] - pd.DateOffset(years=1) + pd.DateOffset(days=1)  # type: ignore
+    )
+
+    # check for missing timestamps and
```
Comment (kvantricht) [inline presto/inference.py]: Do we have a test for this functionality?

File: presto/masking.py [hunk 26]
```diff
@@ -66,7 +66,8 @@ def random_masking(mask, num_tokens_to_mask: int):
             mask[:, SRTM_INDEX] = True
             # then, we flatten the mask and dw arrays
             all_tokens_mask = mask.flatten()
-            unmasked_tokens = all_tokens_mask == False
+            unmasked_tokens = all_tokens_mask == 0
```
Comment (kvantricht) [inline presto/masking.py:73]: This is verified?

File: presto/presto.py [hunk 27]
```diff
@@ -285,11 +283,17 @@ def __init__(
         mlp_ratio=2,
         num_heads=8,
         max_sequence_length=24,
+        valid_month_as_token: bool = False,
```
Comment (kvantricht) [inline presto/presto.py:286]: if it's true, presto expects this input. How does presto understand such input as missing? No data value? a zero? Does it have a mask?

File: train.py [hunk 28]
```diff
@@ -110,323 +137,163 @@
...
+dekadal = False
+compositing_window = "30D"
+if "10d" in parquet_file:
+    dekadal = True
+    compositing_window = "10D"
 
-parquet_file: str = args["parquet_file"]
-val_samples_file: str = args["val_samples_file"]
+valid_month_as_token = False
+if time_token != "none":
```
Comment (kvantricht) [inline train.py]: is `time_token` sufficiently transparent in its meaning? I wouldn't expect it to be linked to `valid_month`.

File: train.py [hunk 29]
```diff
@@ -110,323 +137,163 @@
...
-
-if best_model_path is not None:
-    logger.info("Loading best model: %s" % best_model_path)
-    best_model = torch.load(best_model_path, map_location=device)
-    model.load_state_dict(best_model)
-else:
-    logger.info("Running eval with randomly init weights")
+if task_type == "croptype":
+    df = WorldCerealBase.map_croptypes(df, finetune_classes, downstream_classes)
+    df = filter_remove_noncrops(df)
```
Comment (kvantricht) [inline train.py]: see earlier comment. I think we kind of weaken the crop type classifier here because we only focus on `11`. Shouldn't we also add crops from the `12` class and maybe also some others, to make the crop set even more diverse? I think at some point I was even including grass in order for the crop type model to still be able to work with errors coming out o the crop/no-crop classifier. TBD!

File: train_finetuned.py [hunk 30]
```diff
@@ -0,0 +1,304 @@
...
+argparser.add_argument(
+    "--task_type", type=str, default="croptype", choices=["cropland", "croptype"]
+)
+argparser.add_argument(
+    "--test_type",
+    type=str,
+    default="random",
+    choices=["random", "spatial", "temporal", "seasonal"],
+)
+argparser.add_argument("--time_token", type=str, default="month", choices=["month", "none"])
```
Comment (kvantricht) [inline train_finetuned.py:73]: Change into boolean?
